File: parser.py
# parser.py

from tokenizer import tokenize

from tokenizer import tokenize

class Parser:
    def __init__(self, tokens):
        self.tokens = tokens
        self.position = 0
        self.ast = []

    def current(self):
        if self.position < len(self.tokens):
            return self.tokens[self.position]
        return ('EOF', '')

    def eat(self, expected_type):
        tok_type, tok_val = self.current()
        if tok_type == expected_type:
            self.position += 1
            return tok_val
        raise SyntaxError(f'Expected {expected_type}, got {tok_type}({tok_val})')

    def parse(self):
        while self.current()[0] != 'EOF':
            if self.current()[0] == 'EMIT':
                self.ast.append(self.parse_emit())
            else:
                self.ast.append(self.parse_statement())
        return self.ast

    def parse_expression(self):
        left_type, left_value = self.current()
        if left_type not in ('STRING', 'TEMPLATE', 'NUMBER', 'BOOLEAN', 'IDENT'):
            raise SyntaxError(f'Unexpected value {left_type}')
        self.position += 1

        if self.current()[0] in ('PLUS', 'MINUS', 'STAR', 'SLASH'):
            op_type = self.eat(self.current()[0])
            right_type, right_value = self.current()
            if right_type not in ('STRING', 'TEMPLATE', 'NUMBER', 'BOOLEAN', 'IDENT'):
                raise SyntaxError(f'Expected value after operator, got {right_type}')
            self.position += 1
            return {
                'type': 'binary_expression',
                'operator': op_type,
                'left': (left_type, left_value),
                'right': (right_type, right_value)
            }

        return (left_type, left_value)

    def parse_emit(self):
        self.eat('EMIT')
        value_type, value = self.current()
        if value_type not in ('STRING', 'TEMPLATE', 'IDENT'):
            raise SyntaxError(f'Expected STRING, TEMPLATE, or IDENT after emit, got {value_type}')
        self.position += 1
        self.eat('SEMICOLON')
        return {
            'kind': 'emit',
            'value': (value_type, value)
        }

    def parse_statement(self):
        var_type = self.eat('TYPE')
        var_name = self.eat('IDENT')
        self.eat('EQUALS')
        value = self.parse_expression()
        is_const = False
        if self.current()[0] == 'AS':
            self.eat('AS')
            self.eat('CONST')
            is_const = True
        self.eat('SEMICOLON')
        return {
            'type': var_type,
            'name': var_name,
            'value': value,
            'const': is_const
        }

File: parser.js
// parser.js

export class Parser {
    constructor(tokens) {
        this.tokens = tokens;
        this.position = 0;
        this.ast = [];
    }

    current() {
        return this.position < this.tokens.length
            ? this.tokens[this.position]
            : ['EOF', ''];
    }

    eat(expectedType) {
        const [tokType, tokVal] = this.current();
        if (tokType === expectedType) {
            this.position++;
            return tokVal;
        }
        throw new SyntaxError(`Expected ${expectedType}, got ${tokType}(${tokVal})`);
    }

    parse() {
        while (this.current()[0] !== 'EOF') {
            if (this.current()[0] === 'EMIT') {
                this.ast.push(this.parseEmit());
            } else {
                this.ast.push(this.parseStatement());
            }
        }
        return this.ast;
    }

    parseExpression() {
        const [leftType, leftValue] = this.current();
        if (!['STRING', 'TEMPLATE', 'NUMBER', 'BOOLEAN', 'IDENT'].includes(leftType)) {
            throw new SyntaxError(`Unexpected value ${leftType}`);
        }
        this.position++;

        if (['PLUS', 'MINUS', 'STAR', 'SLASH'].includes(this.current()[0])) {
            const opType = this.eat(this.current()[0]);
            const [rightType, rightValue] = this.current();
            if (!['STRING', 'TEMPLATE', 'NUMBER', 'BOOLEAN', 'IDENT'].includes(rightType)) {
                throw new SyntaxError(`Expected value after operator, got ${rightType}`);
            }
            this.position++;
            return {
                type: 'binary_expression',
                operator: opType,
                left: [leftType, leftValue],
                right: [rightType, rightValue]
            };
        }

        return [leftType, leftValue];
    }

    parseEmit() {
        this.eat('EMIT');
        const [valueType, value] = this.current();
        if (!['STRING', 'TEMPLATE', 'IDENT'].includes(valueType)) {
            throw new SyntaxError(`Expected STRING, TEMPLATE, or IDENT after emit, got ${valueType}`);
        }
        this.position++;
        this.eat('SEMICOLON');
        return {
            kind: 'emit',
            value: [valueType, value]
        };
    }

    parseStatement() {
        const varType = this.eat('TYPE');
        const varName = this.eat('IDENT');
        this.eat('EQUALS');
        const value = this.parseExpression();
        let isConst = false;
        if (this.current()[0] === 'AS') {
            this.eat('AS');
            this.eat('CONST');
            isConst = true;
        }
        this.eat('SEMICOLON');
        return {
            type: varType,
            name: varName,
            value: value,
            const: isConst
        };
    }
}
